arch/xtensa/core/xtensa-win0.c

```diff
@@ -0,0 +1,157 @@
+/* Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <stdint.h>
+#include <xtensa/corebits.h>
+#include <xtensa/config/core-isa.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/kernel_structs.h>
+#include <ksched.h>
+#include <zephyr/arch/xtensa/xtensa-win0.h>
+#include <_soc_inthandlers.h>
+
+/* These are used as part of a mocking layer to test syscall handling
+ * without a full userspace.  Will be removed.
+ */
+__weak int _mock_priv_stack, _k_syscall_table;
+
+/* FIXME: the interrupt handling below is almost-but-not-quite
+ * identical to the code in asm2 (the difference being that win0
+ * passes a NULL to get_next_switch_handle() as it implements a
+ * partial context save and will set the switch handle itself later.
+ * Work out the parameterization and move it somewhere shared.
+ *
+ * (Also this is using printk and not LOG_ERR(), as logging is not
+ * always configured on all apps (and rarely works early enough for
+ * arch-level code).  Low level error logging is not a pretty
+ * situation in Zephyr.)
+ */
+void z_irq_spurious(const __unused void *arg)
+{
+	int irqs, ie;
+
+	__asm__ volatile("rsr.interrupt %0" : "=r"(irqs));
+	__asm__ volatile("rsr.intenable %0" : "=r"(ie));
+	printk(" ** Spurious INTERRUPT(s) %p, INTENABLE = %p\n",
+	       (void *)irqs, (void *)ie);
+	z_xtensa_fatal_error(K_ERR_SPURIOUS_IRQ, NULL);
+}
+
+#define DEF_INT_C_HANDLER(l)						\
+void *xtensa_int##l##_c(void *interrupted)		   		\
+{									\
+	uint32_t irqs, intenable, m;					\
+	IF_ENABLED(CONFIG_SCHED_THREAD_USAGE, z_sched_usage_stop());	\
+	__asm__ volatile("rsr.interrupt %0" : "=r"(irqs));		\
+	__asm__ volatile("rsr.intenable %0" : "=r"(intenable));		\
+	irqs &= intenable;						\
+	while ((m = _xtensa_handle_one_int##l(irqs))) {			\
+		irqs ^= m;						\
+		__asm__ volatile("wsr.intclear %0" : : "r"(m));		\
```

Comment:
This logic (first handle an interrupt / then clear IRQ bit) does not work very well with edge-triggered interrupts, because and IRQ may be set in the gap between the handler completion and wsr.intclear.

Reply:
That's a valid point; this is actually ancient code forklifted over from the original in asm2, and frankly I don't think any of us gave much thought to edge sensitivity.  Just moving the clear to before the handler can be racy with level triggering though if the device doesn't have its own mask to clear in the handler.   Maybe we could handle this with an assert just to disallow edge interrupts for now.

Really the right thing would be to detect the interrupt type at generation time and put the clear there.  We actually do have code to do this, though it's crufty and needs some rework of its own...

Reply:
I swear I have it in my head somewhere that INTCLEAR is used as a bus acknowledgement (i.e. visible to the device) for at least some classes of interrupts?  In which case you can get the converse problem: a device can see the ACK and trigger another interrupt before then handler had run, and the handler would see one event where the device signaled two.

But maybe that's wrong?  This is just vague memory on my part and you're the expert.  If writing INTCLEAR before the handler is safe in all cases then yeah, we should fix this.

Reply:
OK, maybe the best thing is to file this as a bug ("Edge triggered interrupts are racy on xtensa") and fix it in parallel.